### PR TITLE
first step of modular tokenizer refactor

### PR DESCRIPTION
**Pull request by mosheraboh**
* fixed CICD together with Sagi
* detach pretrained tokenizer from modular tokenizer code.
* rename main objects but also keep old names for backward compatibility 

Next step is to move the code to fuse-med-ml (the pre-trained tokenizers will stay in fuse-drug, or maybe should move to bmfm-core). We also need to re-add some tests that were removed in this PR.

### CODE REVIEW

File: fusedrug/data/tokenizer/modulartokenizer/README.md [hunk 1]
```diff
@@ -68,47 +70,27 @@ Saves the underlying adjusted tokenizers as jsons.
 Returns the input token's corresponding ID.
 ## Use example
 ### Creation:
-An example of creation of a new modular tokenizer from a word-level AA sequence tokenizer and a BPE SMILES tokenizer is found here: [ModularTokenizer creation](test_multi_tokenizer_creation.py#L238)
-
-It uses this config: [tokenizer_config.py](configs/tokenizer_config.yaml). Note: this line [path definition](configs/tokenizer_config.yaml#L3) needs to be changed so that _your_path_ points to cloned fuse-drug parent directory.
-
-Additional tokens that are added to the newly created tokenizer are defined in [special_tokens.py](special_tokens.py) in special_tokens and task_tokens. Any additional tokens related to task definitions and queries need to be added to task_tokens. special_tokens and task_tokens are loaded in [ModularTokenizer creation](test_multi_tokenizer_creation.py#L248) by calling get_special_tokens_dict() and get_additional_tokens(["task"]), respectively.
+A script create new modular tokenizer can be found in [ModularTokenizer creation](scripts/create_multi_tokenizer.py).
```
Comment (YoelShoshan) [inline fusedrug/data/tokenizer/modulartokenizer/README.md:73]: ->script that

File: fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py [hunk 2]
```diff
@@ -97,19 +91,14 @@ def __init__(
...
             # store special tokens in a list to preserve their order:
-            all_special_tokens: List
-            if self.special_tokens_dict is None:
-                all_special_tokens = list([])
-            else:
-                all_special_tokens = list(self.special_tokens_dict.values())
-            if additional_tokens_list is not None:
-                all_special_tokens += additional_tokens_list
+            all_special_tokens = (
+                list(special_tokens_list) if special_tokens_list is not None else []
```
Comment (YoelShoshan) [inline fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py:100]: don't we lose the scenario in which all_special_tokens was `list(self.special_tokens_dict.values())` and then `+= additional_tokens_list` ?

Comment (YoelShoshan) [reply fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py:100]: oh, I see it's an argument that was removed from the signature

File: fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py [hunk 3]
```diff
@@ -992,7 +975,7 @@ def count_unknowns(
                 f"Unexpected type of encoding {type(encoding)}, should be list or Encoding"
             )
         if unk_token is None:
-            unk_token = special_wrap_input(special_tokens["unk_token"])
```
Comment (YoelShoshan) [inline fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py:995]: why do you prefer to hardcode it? can't it be different for different (sub) tokenizers?

File: fusedrug_examples/interaction/drug_target/affinity_prediction/PLM_DTI/configs/config.yaml [hunk 4]
```diff
@@ -34,7 +34,7 @@ model:
   target_featurizer: ProtBertFeaturizer
   # possible values: SimpleCoembedding, others (not tested): GoldmanCPI, SimpleCosine, AffinityCoembedInner, CosineBatchNorm, LSTMCosine, DeepCosine,
   #                  SimpleConcat, SeparateConcat, AffinityEmbedConcat, AffinityConcatLinear
-  model_architecture: SimpleCoembedding
+  model_architecture: deepcoembedding
```
Comment (YoelShoshan) [inline fusedrug_examples/interaction/drug_target/affinity_prediction/PLM_DTI/configs/config.yaml:37]: is this intentional in this PR ?

**Review by YoelShoshan — APPROVED** [review]
LGTM

Ideally @floccinauc should also review

File: fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py (hunk 3 above)
Comment (mosheraboh) [reply fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py:995]: The special_tokens was also hard coded. 
We (I 😅 ) need to add such an argument to encode method (like we do with padding)

File: fusedrug/data/tokenizer/modulartokenizer/README.md (hunk 1 above)
Comment (matanninio) [reply fusedrug/data/tokenizer/modulartokenizer/README.md:73]:
```suggestion
A script for creating new modular tokenizers can be found in [ModularTokenizer creation](scripts/create_multi_tokenizer.py).
```

File: fusedrug/data/tokenizer/modulartokenizer/pretrained_tokenizers/configs/tokenizer_config.yaml [hunk 5]
```diff
@@ -1,5 +1,5 @@
 paths:
-  tokenizers_path: "${oc.env:MY_GIT_REPOS}/fuse-drug/fusedrug/data/tokenizer/modulartokenizer/pretrained_tokenizers/" #tokenizer base work path
+  tokenizers_path: #tokenizer base work path
```
Comment (matanninio) [inline fusedrug/data/tokenizer/modulartokenizer/pretrained_tokenizers/configs/tokenizer_config.yaml:2]: I this legit yaml?  is this correct?

Comment (mosheraboh) [reply fusedrug/data/tokenizer/modulartokenizer/pretrained_tokenizers/configs/tokenizer_config.yaml:2]: I was expecting that someone will set it before running the script.
But maybe it's better to set it to null together with asserting that it's different from None at the beginning of the script. Setting it null will allow to override it in command line.

File: fusedrug_examples/interaction/drug_target/affinity_prediction/PLM_DTI/configs/config.yaml (hunk 4 above)
Comment (mosheraboh) [reply fusedrug_examples/interaction/drug_target/affinity_prediction/PLM_DTI/configs/config.yaml:37]: I'll revert it, we ended up disabling this example cause it failed.

File: fusedrug/data/tokenizer/modulartokenizer/scripts/add_tokenizer_to_multi_tokenizer.py [hunk 6]
```diff
@@ -38,16 +28,7 @@ def main(cfg: DictConfig) -> None:
 
     t_mult = ModularTokenizer.load(path=cfg_raw["data"]["tokenizer"]["in_path"])
 
-    test_tokenizer(t_mult, cfg_raw=cfg_raw, mode="loaded_path")
```
Comment (matanninio) [inline fusedrug/data/tokenizer/modulartokenizer/scripts/add_tokenizer_to_multi_tokenizer.py:41]: This is a sanity check for the tokenizer, intended to catch problems before they hit hard.  Why remove this?

**Review by matanninio — COMMENTED** [review]
not very comfortable with dropping the tests.  
If I got this right, the new way of adding special tokens does not allow to create the tokenizer again from the original tokenizers, but requires that the extra tokens be added in just the right order, or one will need to inject the special token list into the first tokenizer.

File: fusedrug/data/tokenizer/modulartokenizer/scripts/add_tokenizer_to_multi_tokenizer.py (hunk 6 above)
Comment (mosheraboh) [reply fusedrug/data/tokenizer/modulartokenizer/scripts/add_tokenizer_to_multi_tokenizer.py:41]: Cause it's should be a generic test here, and this specific test is not in the right place.
Foe example, the script will not necessarily add a subtokenizer to modular tokenizer that already contains AA subtokenizer.

**Review by floccinauc — APPROVED** [review]
Looks great!

File: fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py [hunk 7]
```diff
@@ -1434,15 +1418,7 @@ def update_vocab(
             # operations on the tokenizer instance (if possible, operations should be done here, using built-in tokenizer methods)
             json_str = json.dumps(t_json)
             tokenizer_inst = Tokenizer.from_str(json_str)
-            if self.special_tokens_dict is not None:
```
[the remark this answers is not in the document]
Comment (mosheraboh) [reply fusedrug/data/tokenizer/modulartokenizer/modular_tokenizer.py:1437]: Got it, will add it back
